pkg/clients/dclient/discovery.go

```diff
@@ -92,88 +93,280 @@ func (c serverPreferredResources) GetGVRFromAPIVersionKind(apiVersion string, ki
+		if err != nil {
+			logger.Error(err, "Unable to find parent resource", "kind", kind)
+			return nil, nil, schema.GroupVersionResource{}, err
+		}
+		parentResourceName := parentApiResource.Name
+		resource, gvr, err := findSubresource(groupVersion, parentResourceName, subresource, kind, serverGroupsAndResources)
+		return resource, parentApiResource, gvr, err
 	}
 
-	for _, serverResource := range serverResources {
-		if apiVersion != "" && serverResource.GroupVersion != apiVersion {
-			continue
+	return findResource(groupVersion, kind, serverPreferredResources, serverGroupsAndResources)
+}
+
+// findSubresource finds the subresource for the given parent resource, groupVersion and serverResourcesList
+func findSubresource(groupVersion, parentResourceName, subresource, kind string, serverResourcesList []*metav1.APIResourceList) (apiResource *metav1.APIResource, gvr schema.GroupVersionResource, err error) {
+	for _, serverResourceList := range serverResourcesList {
+		if groupVersion == "" || kubeutils.GroupVersionMatches(groupVersion, serverResourceList.GroupVersion) {
+			for _, serverResource := range serverResourceList.APIResources {
+				if serverResource.Name == parentResourceName+"/"+strings.ToLower(subresource) {
+					logger.V(4).Info("matched API resource to kind", "apiResource", serverResource, "kind", kind)
+
+					serverResourceGv := getServerResourceGroupVersion(serverResourceList.GroupVersion, serverResource.Group, serverResource.Version)
+					gv, err := schema.ParseGroupVersion(serverResourceGv)
+					if err != nil {
+						logger.Error(err, "failed to parse GV", "groupVersion", serverResourceGv)
+						return nil, schema.GroupVersionResource{}, err
+					}
+
+					serverResource.Group = gv.Group
+					serverResource.Version = gv.Version
+
+					groupVersionResource := gv.WithResource(serverResource.Name)
+					logger.V(4).Info("gv with resource", "gvWithResource", groupVersionResource)
+					return &serverResource, groupVersionResource, nil
+				}
+			}
 		}
+	}
 
-		for _, resource := range serverResource.APIResources {
-			if resourceMatches(resource, kind, subresource) {
-				logger.V(4).Info("matched API resource to kind", "apiResource", resource, "kind", kind)
-				gv, err := schema.ParseGroupVersion(serverResource.GroupVersion)
-				if err != nil {
-					logger.Error(err, "failed to parse GV", "groupVersion", serverResource.GroupVersion)
-					return nil, schema.GroupVersionResource{}, err
+	return nil, schema.GroupVersionResource{}, fmt.Errorf("resource not found for kind %s", kind)
+}
+
+// findResource finds an API resource that matches 'groupVersion', 'kind', in the given serverResourcesList
+func findResource(groupVersion string, kind string, serverPreferredResources, serverGroupsAndResources []*metav1.APIResourceList) (apiResource, parentAPIResource *metav1.APIResource, gvr schema.GroupVersionResource, err error) {
+
+	type APIResourceWithListGV struct {
```

Comment:
Do we need to expose `APIResourceWithListGV`?

Reply:
Nope, fixed.